node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/v2/filter/AuthorizationFilter.java

```diff
@@ -0,0 +1,104 @@
+// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.hosted.provision.restapi.v2.filter;
+
+import com.google.inject.Inject;
+import com.yahoo.config.provision.Zone;
+import com.yahoo.jdisc.handler.FastContentWriter;
+import com.yahoo.jdisc.handler.ResponseDispatch;
+import com.yahoo.jdisc.handler.ResponseHandler;
+import com.yahoo.jdisc.http.filter.DiscFilterRequest;
+import com.yahoo.jdisc.http.filter.SecurityRequestFilter;
+import com.yahoo.jdisc.http.servlet.ServletRequest;
+import com.yahoo.vespa.hosted.provision.NodeRepository;
+import com.yahoo.vespa.hosted.provision.restapi.v2.Authorizer;
+import com.yahoo.vespa.hosted.provision.restapi.v2.ErrorResponse;
+import org.bouncycastle.asn1.x500.RDN;
+import org.bouncycastle.asn1.x500.X500Name;
+import org.bouncycastle.asn1.x500.style.BCStyle;
+import org.bouncycastle.asn1.x500.style.IETFUtils;
+import org.bouncycastle.cert.jcajce.JcaX509CertificateHolder;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.URI;
+import java.security.Principal;
+import java.security.cert.CertificateEncodingException;
+import java.security.cert.X509Certificate;
+import java.util.Optional;
+import java.util.function.BiConsumer;
+import java.util.function.BiPredicate;
+import java.util.logging.Logger;
+
+/**
+ * @author mpolden
+ */
+public class AuthorizationFilter implements SecurityRequestFilter {
+
+    private static final Logger log = Logger.getLogger(AuthorizationFilter.class.getName());
+
+    private final BiPredicate<Principal, URI> authorizer;
+    private final BiConsumer<ErrorResponse, ResponseHandler> responseWriter;
+
+    @Inject
+    public AuthorizationFilter(Zone zone, NodeRepository nodeRepository) {
+         this(new Authorizer(zone.system(), nodeRepository), AuthorizationFilter::log); // TODO: Use write method once all clients are using certificates
+    }
+
+    AuthorizationFilter(Authorizer authorizer, BiConsumer<ErrorResponse, ResponseHandler> responseWriter) {
+        this.authorizer = authorizer;
+        this.responseWriter = responseWriter;
+    }
+
+    @Override
+    public void filter(DiscFilterRequest request, ResponseHandler handler) {
+        Optional<X509Certificate> cert = certificateFrom(request);
+        if (!cert.isPresent() || !authorizer.test(() -> commonName(cert.get()), request.getUri())) {
```

Comment:
The filter should distinguish 403 Forbidden (authorization failure) from 401 Unauthorized (authentication failure). Missing credentials => 401, invalid credentials => 403. See https://stackoverflow.com/a/6937030 :)